source/features/render-semantic-types.tsx

```diff
@@ -0,0 +1,191 @@
+import React from 'react';
+import {$$} from 'select-dom';
+import * as pageDetect from 'github-url-detection';
+
+import features from '../feature-manager.js';
+
+function isSemanticCommitTitleElement(commitTitle: string): boolean {
+	// This may not be fully accurate.
+	// Tests, if the title string
+	// starts with a word,
+	// followed by an anything in parentheses,
+	// followed by an optional exclamation mark,
+	// followed by a colon,
+	// followed by anything.
+	return /^\w*(?:\(.*\))?!?:.*/.test(commitTitle);
+}
+
+const defaultCommitTypes = new Set([
+	'feat',
+	'fix',
+	'chore',
+	'docs',
+	'build',
+	'refactor',
+	'test',
+	'ci',
+	'perf',
+]);
+
+function isSemanticCommitTitleTypeDefault(semanticCommitTitle: string): boolean {
+	// This may not be fully accurate.
+	// Gets the first word from the title.
+	const type = /^(\w*)/.exec(semanticCommitTitle)?.pop();
+
+	if (!type) {
+		return false;
+	}
+
+	return defaultCommitTypes.has(type);
+}
+
+function extractSemanticCommitTitleTypeAndScope(semanticCommitTitle: string): [string, string?] {
+	// Gets the first word from the title.
+	const type = /^(\w*)/.exec(semanticCommitTitle)!.pop()!;
+
+	// Gets the word in parentheses from the title.
+	const scope = /\((\w*)\)/.exec(semanticCommitTitle)?.pop();
+
+	return [type, scope];
+}
+
+function removeSemanticCommitTitleAndScopeFromTitleElement(semanticCommitTitleElement: HTMLElement): void {
+	const children = semanticCommitTitleElement.childNodes;
+
+	// Remove all children until the child with a colon is found.
+	// Then remove the colon and everything before it.
+	for (const child of children) {
+		if (child.textContent.includes(':')) {
+			child.textContent = child.textContent.split(':').slice(1).join(':');
```

Comment:
Like said above, we should try to keep the logic in a single place, so ideally we’d still use the regex to remove the same exact part of the title. This however isn’t straightforward across inline elements, it might need to be done via Range API.

I think however this can also be easily solved via zip-text-nodes. We can review it later.

Reply:
Here’s some code written with Claude:

``` JavaScript 
function removeText(element, searchRegex) {
  const modifiedText = element.textContent.replace(searchRegex, match => `<s>${match}</s>`);
  
  const temp = document.createElement('div');
  temp.innerHTML = modifiedText;

  zipTextNodes(element, temp);

  element.querySelector('s').remove();
}
```

I should do the job.

So the logic would probably be like:

``` JavaScript 
const match = regex.exec(commitTitle.textContent)

if (!match) return

const {type, scope} = match.groups;
prependLabel(commitTitle, type, scope)
removeText(commitTitle, match[0])
```

I think all these changes would make the feature really small actually

Reply:
This solution seems less performant. I prefer my current solution because it doesn't rely on libraries or add any temporary div to a DOM. zipTextNodes feels like a hack on it's own as well. I applied your suggestion anyways, but I'd suggest reverting it.